### PR TITLE
Fix memory leaks in pkread.c demo file

### LINKED ISSUE #8456: Many memory leaks reported by valgrind in PKCS12_parse
Hi,

I used demos/pkread.c to read in a pkcs12 file test/shibboleth.pfx. While it works correctly, valgrind reports a lot of memory leaks. This can be reproduced using any pfx file and using any test program with PKCS12_parse.

I could not find the root cause of these leaks. Could someone take a look? Thanks for your time.

==119055== HEAP SUMMARY:
==119055==     in use at exit: 7,465 bytes in 84 blocks
==119055==   total heap usage: 3,366 allocs, 3,282 frees, 2,258,849 bytes allocated
==119055== 
==119055== 37 bytes in 1 blocks are definitely lost in loss record 14 of 47
==119055==    at 0x4C2BBAF: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==119055==    by 0x4FF444D: CRYPTO_malloc (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x5016645: OPENSSL_uni2utf8 (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x5012624: PKCS12_get_friendlyname (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x400E31: find_friendly_name (pkread.c:38)
==119055==    by 0x40101C: main (pkread.c:80)
==119055== 
==119055== 1,712 (72 direct, 1,640 indirect) bytes in 1 blocks are definitely lost in loss record 44 of 47
==119055==    at 0x4C2BBAF: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==119055==    by 0x4FF444D: CRYPTO_malloc (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x4FF4480: CRYPTO_zalloc (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x4FE4A49: EVP_PKEY_new (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x4FE15D1: EVP_PKCS82PKEY (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x50146B7: parse_bag (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x50144F7: parse_bags (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x501441F: parse_pk12 (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x5014181: PKCS12_parse (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x400FD5: main (pkread.c:75)
==119055== 
==119055== 2,572 (32 direct, 2,540 indirect) bytes in 1 blocks are definitely lost in loss record 46 of 47
==119055==    at 0x4C2BBAF: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==119055==    by 0x4FF444D: CRYPTO_malloc (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x4FF4480: CRYPTO_zalloc (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x506A614: OPENSSL_sk_new_reserve (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x506A3BD: OPENSSL_sk_new_null (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x4EE1226: sk_ASN1_VALUE_new_null (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x4EE24BB: asn1_template_noexp_d2i (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x4EE2338: asn1_template_ex_d2i (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x4EE151C: asn1_item_embed_d2i (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x4EE1365: ASN1_item_ex_d2i (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x4EE12F9: ASN1_item_d2i (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x4EDCD22: ASN1_item_unpack (in /openssl/lib/libcrypto.so.1.1)
==119055== 
==119055== 3,144 (352 direct, 2,792 indirect) bytes in 1 blocks are definitely lost in loss record 47 of 47
==119055==    at 0x4C2BBAF: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==119055==    by 0x4FF444D: CRYPTO_malloc (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x4FF4480: CRYPTO_zalloc (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x4EE550A: asn1_item_embed_new (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x4EE5260: ASN1_item_ex_new (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x4EE1B9F: asn1_item_embed_d2i (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x4EE1365: ASN1_item_ex_d2i (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x4EE12F9: ASN1_item_d2i (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x4EDCD22: ASN1_item_unpack (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x5015DC0: PKCS12_SAFEBAG_get1_cert (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x5014721: parse_bag (in /openssl/lib/libcrypto.so.1.1)
==119055==    by 0x50144F7: parse_bags (in /openssl/lib/libcrypto.so.1.1)
==119055== 
==119055== LEAK SUMMARY:
==119055==    definitely lost: 493 bytes in 4 blocks
==119055==    indirectly lost: 6,972 bytes in 80 blocks
==119055==      possibly lost: 0 bytes in 0 blocks
==119055==    still reachable: 0 bytes in 0 blocks
==119055==         suppressed: 0 bytes in 0 blocks
==119055==

### PR DESCRIPTION
**Pull request by mattcaswell**
Also make various changes to bring the file into line with current coding
style.

Fixes #8456

### CODE REVIEW

File: demos/pkcs12/pkread.c [hunk 1]
```diff
@@ -48,56 +48,64 @@ static char *find_friendly_name(PKCS12 *p12)
...
-    EVP_PKEY *pkey;
-    X509 *cert;
+    EVP_PKEY *pkey = NULL;
+    X509 *cert = NULL;
     STACK_OF(X509) *ca = NULL;
-    PKCS12 *p12;
-    const char *name;
-    int i;
+    PKCS12 *p12 = NULL;
+    char *name;
```
Comment (InfoHunter) [inline demos/pkcs12/pkread.c]: Should this be set to NULL? (`char *name = NULL;`)

Comment (InfoHunter) [reply demos/pkcs12/pkread.c]: It seems the `OPENSSL_free` in the `err` section could free this pointer before setting a value to it.

Comment (paulidale) [reply demos/pkcs12/pkread.c]: It look this way to me too

File: demos/pkcs12/pkread.c [hunk 2]
```diff
@@ -48,56 +48,64 @@ static char *find_friendly_name(PKCS12 *p12)
...
+    int i, ret = 1;
```
Comment (paulidale) [inline demos/pkcs12/pkread.c]: This should be EXIT_FAILURE instead of 1.

File: demos/pkcs12/pkread.c [hunk 3]
```diff
@@ -48,56 +48,64 @@ static char *find_friendly_name(PKCS12 *p12)
...
 
     if (argc != 4) {
         fprintf(stderr, "Usage: pkread p12file password opfile\n");
         exit(1);
```
Comment (paulidale) [inline demos/pkcs12/pkread.c]: exit(ret) or return ret or goto err here

File: demos/pkcs12/pkread.c [hunk 4]
```diff
@@ -48,56 +48,64 @@ static char *find_friendly_name(PKCS12 *p12)
...
     }
-    OpenSSL_add_all_algorithms();
-    ERR_load_crypto_strings();
+
     if ((fp = fopen(argv[1], "rb")) == NULL) {
         fprintf(stderr, "Error opening file %s\n", argv[1]);
         exit(1);
```
Comment (paulidale) [inline demos/pkcs12/pkread.c]: Likewise

File: demos/pkcs12/pkread.c [hunk 5]
```diff
@@ -48,56 +48,64 @@ static char *find_friendly_name(PKCS12 *p12)
...
-    if (ca && sk_X509_num(ca)) {
+    if (ca != NULL && sk_X509_num(ca) > 0) {
         fprintf(fp, "***Other Certificates***\n");
         for (i = 0; i < sk_X509_num(ca); i++)
             PEM_write_X509_AUX(fp, sk_X509_value(ca, i));
     }
     fclose(fp);
-    return 0;
+
+    ret = 0;
```
Comment (paulidale) [inline demos/pkcs12/pkread.c]: ret = EXIT_SUCCESS here

**Comment by mattcaswell** [comment]
Fixup commits pushed addressing all feedback.

File: demos/pkcs12/pkread.c (hunk 3 above)
Comment (paulidale) [reply demos/pkcs12/pkread.c]: My pick would be _goto err_ but it’s not a big deal.

**Comment by paulidale** [comment]
The demos could do with a going over at some point. All the ones I’ve looked at needed cleaning like this. Descriptive comments would also be helpful.

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**